test/testers.jl

```diff
@@ -42,62 +42,42 @@ primalapprox(x) = x
         end
     end
 
-    @testset "Inplace accumumulation: first on Array" begin
+    @testset "Inplace accumulation: identity on Array" begin
         @testset "Correct definitions" begin
-            function ChainRulesCore.frule((_, ẋ), ::typeof(first), x::Array)
-                ẏ = InplaceableThunk(
-                    @thunk(first(ẋ)),
-                    ȧ -> ȧ + first(ẋ),  # This won't actually happen inplace
-                )
-                return first(x), ẏ
+            function ChainRulesCore.frule((_, ẋ), ::typeof(identity), x::Array)
+                ẏ = InplaceableThunk(@thunk(ẋ), ȧ -> ȧ .+= ẋ)
+                return identity(x), ẏ
             end
-            function ChainRulesCore.rrule(::typeof(first), x::Array{T}) where T
-                x_dims = size(x)
-                function first_pullback(ȳ)
-                    x̄_ret = InplaceableThunk(
-                        Thunk() do
-                            x̄ = zeros(T, x_dims)
-                            x̄[1]=ȳ
-                            x̄
-                        end,
-                        ā -> (ā[1] += ȳ; ā)
-                    )
+            function ChainRulesCore.rrule(::typeof(identity), x::Array)
+                function identity_pullback(ȳ)
+                    x̄_ret = InplaceableThunk(@thunk(ȳ), ā -> ā .+= ȳ)
                     return (NO_FIELDS, x̄_ret)
                 end
-                return first(x), first_pullback
+                return identity(x), identity_pullback
             end
 
-            frule_test(first, (randn(4), randn(4)))
-            rrule_test(first, randn(), (randn(4), randn(4)))
+            frule_test(identity, (randn(4), randn(4)))
+            rrule_test(identity, randn(4), (randn(4), randn(4)))
```

Comment:
I am not sure I understand what this test was meant to catch before (and therefore how the changes improve it).

Was it just a general test of inplace accummulation using `InplaceableThunk`s? If so, does the new `add!!` function just use the normal `x+y` rather than the inplace computation?

Reply:
Yes, just a general trest of inplace accumulation using `InplacableThunk`.
But the new `add!!` is too smart and will do `x+y` instread of the inplace computation,
thus defeating the point of this test.
It's not a great test in that it can be silently defeated.
Ideally I guess we would intrument it and check that the code paths we want to be hit are hit.
I think i can workout how to do that.